### PR TITLE
Use 48-bit commitment transaction numbers

### PR DESCRIPTION
**Pull request by cjamthagen**
TODO: fix deriveStateHintObsfucator to set correct obsfucator for the channel. According to the [spec](https://github.com/lightningnetwork/lightning-rfc/blob/master/03-transactions.md#commitment-transaction) the payment-basepoints from `open_channel` and `accept_channel` are needed, and they don't seem to be implemented yet.

Since I am pretty new to writing Go, please let me know if I have done something wrong.

### CODE REVIEW

File: lnwallet/script_utils.go [hunk 1]
```diff
@@ -790,16 +791,20 @@ func SetStateNumHint(commitTx *wire.MsgTx, stateNum uint32,
 			"instead has %v", len(commitTx.TxIn))
 	}
 
-	// Convert the obfuscater into a uint32, then XOR that against the
+	// Convert the obfuscator into a uint64, then XOR that against the
 	// targeted height in order to obfuscate the state number of the
 	// commitment transaction in the case that either commitment
 	// transaction is broadcast directly on chain.
-	xorInt := binary.BigEndian.Uint32(obsfucator[:]) & (^wire.SequenceLockTimeDisabled)
+	obfs := make([]byte, 8)
```
Comment (Roasbeef) [inline lnwallet/script_utils.go]: We can avoid a heap allocation here by declaring the slice as a fixed-sized array like so: 
```
var obfs [8]byte
```

File: lnwallet/script_utils.go [hunk 2]
```diff
@@ -810,14 +815,17 @@ func SetStateNumHint(commitTx *wire.MsgTx, stateNum uint32,
...
-	// Convert the obfuscater into a uint32, this will be used to
+func GetStateNumHint(commitTx *wire.MsgTx, obsfucator [StateHintSize]byte) uint64 {
+	// Convert the obfuscater into a uint64, this will be used to
 	// de-obfuscate the final recovered state number.
-	xorInt := binary.BigEndian.Uint32(obsfucator[:]) & (^wire.SequenceLockTimeDisabled)
-
-	// Retrieve the sole state hint from the sequence number of the
-	// transaction. In the process un-set the top bit.
-	stateNumXor := commitTx.TxIn[0].Sequence & (^wire.SequenceLockTimeDisabled)
+	obfs := make([]byte, 8)
```
Comment (Roasbeef) [inline lnwallet/script_utils.go]: Same comment here as above about using a fixed-sized array instead of a dynamically allocated slice.

**Comment by cjamthagen** [comment]
Fixed.

Question: Is there a specific way to rebase to master that is preferred?

Will work on the tests next.

**Comment by Roasbeef** [comment]
Great, thanks! 

I've answered your question about rebasing in [this comment on your other PR](https://github.com/lightningnetwork/lnd/pull/92#issuecomment-270795527).

**Comment by cjamthagen** [comment]
Table driven tests added.

**Comment by cjamthagen** [comment]
It seems like the locktime is the culprit for not passing the 'channel force closure' integration test.

The lower 24 bits of the locktime represents the lower 24 bits of the obscured commitment transaction number. If the high byte of locktime is always zero, then the locktime value can only assume a maximum value of 16,777,216, which is below 500,000,000, meaning the locktime will always be compared to blockheight and not MTP, if enabled. Similarly the sequence number will never take the value maxInt (0xFFFFFFFF) because the high byte will always be 0x80. This means that the locktime will not be disabled for the commit transaction, and the commit transaction will not be eligible for inclusion in a block until it has the proper height as constrained by its locktime.

If this is the case, then the locktime cannot be used as defined in the specification to hold the obscured commitment transaction number, because it would risk rendering the commitment transaction unspendable in many cases, only passing as spendable in 2-3% of the time on the mainchain.

Let me know I got anything wrong.

**Comment by cjamthagen** [comment]
Alright, setting the 30th bit (1<<29) of the locktime fixes the problem and allows the test to pass. I'll await some comments and then propose an update to BOLT 3.

**Comment by cjamthagen** [comment]
Fixed. I'll take a look at the test vectors in the spec and make sure they get added.

**Comment by cjamthagen** [comment]
Done. What `ChannelDelta` usage needed modification?

**Comment by Roasbeef** [comment]
@cjamthagen this modification: https://github.com/lightningnetwork/lnd/commit/db8c6d4cf4b761e911f57931fefbede89097cb16

**Pull request merged**